# Probability Theory

### Probability vs Statistics

**Probability** - predictions about future events based on models or causes that we assume; ***predicting data***  
**Statistics** - analyze data and past events to infer what those models or causes could be; ***using data to predict***

##### Basic Probability

* Probability of Event = $P$
* Probability of a Complimentary Event = $1-P$
* Probability of Composite Independent Events = $\prod{P}$
* The probability of any event must be between $[0, 1]$.

##### Example - Coin Flips

$P(\text{H}) = 0.5$  
$P(\text{T}) = 1 - P(\text{H}) = P(\text{not H}) = 0.5$

Across $n$ coin flips, the probability of seeing $n$ heads is $P(\text{H})^n$.

* Probability of Event = $P$
* Probability of a Complimentary Event = $1-P$
* Probability of Composite Independent Events = $\prod{P}$
* The probability of any event must be between $[0, 1]$.

##### Example - Coin Flips

$P(\text{H}) = 0.5$  
$P(\text{T}) = 1 - P(\text{H}) = P(\text{not H}) = 0.5$

Across $n$ coin flips, the probability of seeing $n$ heads is $P(\text{H})^n$.

In [30]:
import math
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from timeit import default_timer as timer
%matplotlib inline

##### Flipping a Coin

In [31]:
np.random.randint(2, size=10000).mean()

0.5043

In [32]:
np.random.choice([0, 1], size=10000, p=[0.8, 0.2]).mean()

0.2062

##### Two Fair Coin Flips - Percentage of Two Heads?

In [33]:
# 0 is Heads, 1 is Tails
tests = np.random.randint(2, size=(int(1e6), 2))
print(tests[:10])
(tests.sum(axis=1) == 0).mean()

[[1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 0]
 [1 0]]


0.24984

##### Three Fair Coin Flips - Probability Exactly One Head?

In [34]:
tests = np.random.randint(2, size=(int(1e6), 3))
(tests.sum(axis=1) == 2).mean()

0.374764

##### Three Biased Coin Flips - $P(\text{H}) = 0.6 $ - Probability Exactly One Head?

In [35]:
tests = np.random.choice([0,1], size=(int(1e6), 3), p=[0.6, 0.4])
(tests.sum(axis=1) == 2).mean()

0.288142

##### Die Roll - Probability of an Even Number?

In [36]:
tests = np.random.randint(6, size=int(1e6))
(tests%2 == 0).mean()

0.499699

##### Two Dice Rolls - Probability of a Double?

In [37]:
start = timer()

tests = np.random.randint(6, size=(int(1e6), 2))
print([np.all(row == row[0]) for row in tests].count(True) / len(tests))
timer() - start

0.166822


5.5248169240003335

In [38]:
start = timer()

tests1 = np.random.randint(6, size=int(1e6))
tests2 = np.random.randint(6, size=int(1e6))

(tests1 == tests2).mean()

timer() - start

0.019433597999523045

### Counting

If the order of selection of $n$ items being counted $k$ at a time matters, then the method for counting possible **permutations** is $$^nP_k = n*(n-1)*\dots*(n-k+1)=\frac{n!}{(n-k)!}$$

If the order of selection does *not* matter, then the method for counting possible **combinations** is $${n \choose k} = \frac{n!}{k!(n-k)!}$$

##### Permutation Examples

1. How many words can be formed using the letters of the word "TRIANGLES"?

With 9 distinct letters, the number of different permutations is $^9P_9 = 9!$

2. What if we fix T at the beginning and S at the end?

We then have 7 letters we can permute. This can be done in $^7P_7$ ways.

3. How many different 5-letter words can be formed using the letters from A to J (10 letters) such that each word has at least one letter repeated.

First, we compute the number of possible 5-letter words (with or without replacement) using 10 distinct letters, which is $10^5 = 10*10*10*10*10$ as we can always choose 10 letters for a slot.  
Next we compute the number of possible 5-letter words without repetition. This is $^{10}P_5$.  
Thus, the the answer is the difference between the set of all possible 5-letter words minus the 5-letter words without repetition. This leaves only the 5-letter words with at least one repetition. $10^5- ^{10}P_5$

4. There are 10 students in a class. Two of these students, Jack and Daniel, are not permitted to sit together. How many ways can the teacher arrange the students?

The total number of arrangement of students is $^{10}P_{10} = 10!$.  
The number of arrangements where Jack and Daniel are together can be computed by treating them as a single unit. Thus, these 9 entities can be arranged in $^9P_9=9!$ ways. However, we must also consider the ways Jack and Daniel themselves can be permuted, which is $2!$ or 2 ways. Thus, the total number of ways in which Jack and Daniel are together is $2*9!$.  
We can conclude that the number of permutations in which Jack and Daniel are not together is $10!-(2*9!)$.

##### Combination Example

1. How many ways can you flip a fair coin 10 times and obtain 4 heads?

${10 \choose 4} = \frac{10!}{4!6!} = 210$

In [39]:
math.factorial(10)/(math.factorial(6)*math.factorial(4))

210.0

### Random Variables

A random variable, $X$ is a quantity with an associated probability distribution. It can either be discrete (i.e., have a countable range) or continuous (i.e., have an uncountable range).

The probability distribution associated with a discrete random variable is a probability mass function (PMF), and that associated with a continuous random variable is a probability density function (PDF). Both can be represented as $f_X(x)$.

Probabilities of both discrete and continuous random variables must be non-negative and sum or integrate to 1.

$$\text{Discrete: }\sum\limits_{x\in X}f_X(x)=1$$
$$\text{Continous: }\int\limits_{-\infty}^{\infty}f_X(x)dx=1$$

The cumulative distribution function (CDF) is often used in practice rather than a variable's PMF or PDF and is defined as $F_X(x) = p(X \leq x)$.
$$\text{Discrete: }F_X(x) = \sum\limits_{k \leq x}p(k)$$
$$\text{Continuous: }F_X(x) = \int\limits_{-\infty}^{x}p(y)dy$$

Thus, the CDR, which is non-negative and monotonically increasing, can be obtained by taking the sum of PDFs in the discrete case or integral of PDFs in the continuous case.

### Conditional Probability

When the outcome of one event depends on an earlier event, we must use conditional probabilities.

$P(A|B) = \frac{P(A \cap B)}{P(B)} = \frac{P(B|A)P(A)}{P(B)}$

where $|$ is read as *given* and $\cap$ represents *and*. So the above would read, "The probability of $A$ given $B$ is the probability of $A$ and $B$ divided by the probability of $B$."

$\text{posterior} = \frac{\text{joint}}{\text{normalization-factor}} = \frac{\text{likelihood}*\text{prior}}{\text{normalization-factor}}$

Likelihood - $P(B|A)$  
Prior - $P(A)$  
Posterior - $P(A|B)$

In the OneNote examples, we rearranged the above to compute

$P(\text{positive} | \text{disease}) = \frac{P(\text{positive} \cap \text{disease})}{P(\text{disease})}$

$P(\text{positive} \cap \text{disease}) = P(\text{positive} | \text{disease})P(\text{disease})$


### Law of Total Probability

For a set of $n$ disjoint events within $B$ having occurred, the $P(A)$ also having occurred can be computed as follows: $P(A) = P(A|b_1)P(b_1)+\dots+P(A|b_n)P(b_n) = P(A, b_1)+\dots+P(A, b_n)=\sum\limits_{b\in B}P_{A,B}(a,b)$.

Thus, by conditioning $A$ on all possible values of $B$, we obtain an intersection that sums to the total area of $P(A)$.

### Independence

If the outcome of $A$ does not depend on $B$, then $P(A|B) = P(A)$.

Similarly, it is possible for $A$ and $B$ to be conditionally independent. For instance, given that $C$ has occurred, knowing that $B$ also occurred tells us nothing about the probability of $A$ having occurred. $P\left(A \cap B|C\right) = P(A|C)P(B|C)$.

### Join, Marginal, and Conditional Probability Distributions

Random variables are often analyzed with respect to other random variables, giving rise to *joint PMFs* in the discrete case and *joint PDFs* in the continuous case.
$$\int\limits_{-\infty}^{\infty}\int\limits_{-\infty}^{\infty}f_{X,Y}(x,y)dxdy=1$$

From a joint PDF or PMF, a marginal PDF or PMF can be derived using the law of total probability. In the case of marginal PDF for $x$, which can be obtained by integrating with respect to $y$.
$$\text{Discrete: }f_X(x)=\sum\limits_{y\in Y}f_{X,Y}(x,y)=\sum\limits_{y\in Y}f_{X|Y}(x|y)f_Y(y)$$
$$\text{Continuous: }f_X(x)=\int\limits_{-\infty}^{\infty}f_{X,Y}(x,y)dy=\int\limits_{-\infty}^{\infty}f_{X|Y}(x|y)f_Y(y)dy$$

Similarly, we can define a joint CDF where $F_{X,Y}(x,y) = P(X\leq x, Y\leq y)$ is equivalent to
$$F_{X,Y}(x,y)=\int\limits_{-\infty}^{x}\int\limits_{-\infty}^{y}f_{X,Y}(u,v)dvdu$$

### Bayes Rule

Imagine we have a variable we care about that is hidden, meaning it cannot be measured directly, but that we do have a test. Let's use the example of cancer and test that tests either positive or negative.

The knowledge we come to the problem with is known as our **prior**. In this case, our prior is the prevalence of cancer, where $C$ is our random variable for cancer.

**Prior**:  $P(\text{C})$

Once we begin testing, we begin to obtain evidence that add information to our prior. In this case, positive ($\text{T}_\text{p}$) and negative ($\text{T}_\text{n}$) test results.

**Sensitivity**: $P(\text{T}_\text{p} | \text{C})$ - true positive rate  
**Specificity**: $P(\text{T}_\text{n} | \lnot\text{C})$ - true negative rate

Using our prior and our evidence, we can compute joint distributions. We will first compute the joint distribution for positive test cases.

$P(\text{C}, \text{T}_\text{p}) = P(\text{T}_\text{p} | \text{C})P(\text{C})$  
$P(\lnot\text{C}, \text{T}_\text{p}) = P(\text{T}_\text{p} | \lnot\text{C})P(\lnot\text{C}) = (1 - P(\text{T}_\text{n}|\lnot\text{C}))(1 - P(\text{C}))$

Once we have our joint distributions, we can compute $P(\text{T}_\text{p})$. Note that the joint distributions are required as the conditional probabilities, $P(\text{T}_\text{p} | \text{C})$ and $P(\text{T}_\text{p} | \lnot\text{C})$ do not consider the prevelance of $C$.

$P(\text{T}_\text{p}) = P(\text{C}, \text{T}_\text{p}) + P(\lnot\text{C}, \text{T}_\text{p}) = P(\text{T}_\text{p} | \text{C})P(\text{C}) + P(\text{T}_\text{p} | \lnot\text{C})P(\lnot\text{C})$

This is known as the **Law of Total Probability** which will be defined below.

We will use $P(\text{T}_\text{p})$ to normalize our joint distributions and compute our posterior probability distributions. We can think of the joint distributions as portions of a total area defined by $P({\text{T}_\text{p}})$. (See OneNote for example drawing.)

$P(\text{C} | \text{T}_\text{p}) = \frac{P(\text{C}, \text{T}_\text{p})}{P(\text{T}_\text{p})}$  
$P(\lnot\text{C} | \text{T}_\text{p}) = \frac{P(\lnot\text{C}, \text{T}_\text{p})}{P(\text{T}_\text{p})}$

To compute the posterior probabilities given negative test results, $P(\text{C}|\text{T}_\text{n})$ and $P(\lnot\text{C}|\text{T}_\text{n})$, we can swap $\text{T}_\text{p}$ and $\text{T}_\text{n}$ in the above algorithm.

### Python Practice

This exercise will use a simulated dataset on cancer test results for patients and whether they really have cancer.

* How many patients are there in total?
* How many patients have cancer?
* How many patients do not have cancer?
* What proportion of patients have cancer?
* What proportion of patients don't have cancer?
* What proportion of patients with cancer test positive?
* What proportion of patients with cancer test negative?
* What proportion of patients without cancer test positive?
* What proportion of patients without cancer test negative?

In [40]:
# Load dataset
df = pd.read_csv('Data/cancer_test_data.csv')
df.head()

patient_id test_result  has_cancer
0       79452    Negative       False
1       81667    Positive        True
2       76297    Negative       False
3       36593    Negative       False
4       53717    Negative       False

In [41]:
# Number of Patients
df.shape[0]

2914

In [42]:
# Number of Patients with Cancer
df.has_cancer.sum()

306

In [43]:
# Number of Patients without Cancer
df.shape[0] - df.has_cancer.sum()

2608

In [44]:
# Proportion of Patients with Cancer, P(C)
df.has_cancer.mean()

0.10501029512697323

In [45]:
# Proportion of Patients without Cancer, P(~C)
1 - df.has_cancer.mean()

0.8949897048730268

In [46]:
# Proportion of Patients with Cancer who tested Positive, P(Pos|C)
(df[df.has_cancer]['test_result'] == "Positive").mean()

0.9052287581699346

In [47]:
# Proportion of Patients with Cancer who tested Negative, P(Neg|C)
(df[df.has_cancer]['test_result'] == "Negative").mean()

0.09477124183006536

In [48]:
# Proportion of Patients without Cancer who tested Positive, P(Pos|~C)
(df[df.has_cancer == False]['test_result'] == "Positive").mean()

0.2036042944785276

In [49]:
# Proportion of Patients without Cancer who tested Negative, P(Neg|~C)
(df[df.has_cancer == False]['test_result'] == "Negative").mean()

0.7963957055214724

##### Results

$P(\text{C}) = 0.105$  
$P(\lnot\text{C}) = 0.895$  
$P(\text{pos}|\text{C}) = 0.905$  
$P(\text{neg}|\text{C}) = 0.095$  
$P(\text{pos}|\lnot\text{C}) = 0.204$  
$P(\text{neg}|\lnot\text{C}) = 0.796$  

Find the following probabilities, the probability of having cancer or not given a positive or negative test result.  
$P(\text{C}|\text{pos})$  
$P(\lnot\text{C}|\text{pos})$  
$P(\text{C}|\text{neg})$  
$P(\lnot\text{C}|\text{neg})$  

1. $P(\text{pos}) = P(\text{C}, \text{pos}) + P(\lnot\text{C}, \text{pos}) = P(\text{pos}|\text{C})P(\text{C}) + P(\text{pos}|\lnot\text{C})P(\lnot\text{C})$ 
1. $P(\text{neg}) = P(\text{C}, \text{neg}) + P(\lnot\text{C}, \text{neg}) = P(\text{neg}|\text{C})P(\text{C}) + P(\text{neg}|\lnot\text{C})P(\lnot\text{C})$ 
1. $P(\text{C}|\text{pos}) = \frac{P(\text{C}, \text{pos})}{P(\text{pos})}$
1. $P(\lnot\text{C}|\text{pos}) = \frac{P(\lnot\text{C}, \text{pos})}{P(\text{pos})} = 1 - P(\text{C}|\text{pos})$
1. $P(\text{C}|\text{neg}) = \frac{P(\text{C}, \text{neg})}{P(\text{neg})}$
1. $P(\lnot\text{C}|\text{neg}) = \frac{P(\lnot\text{C}, \text{neg})}{P(\text{neg})} = 1 - P(\text{C}|\text{neg})$

In [50]:
print('P(pos) using the dataframe directly:', (df['test_result'] == "Positive").mean())
print('P(pos) using Bayes rule (and rounded results):', .905*.105 + .204*.895)

P(pos) using the dataframe directly: 0.2772820864790666
P(pos) using Bayes rule (and rounded results): 0.277605


In [51]:
print('P(neg) using the dataframe directly:', (df['test_result'] == "Negative").mean())
print('P(neg) using Bayes rule (and rounded results):', .095*.105 + .796*.895)

P(neg) using the dataframe directly: 0.7227179135209334
P(neg) using Bayes rule (and rounded results): 0.722395


In [52]:
print('P(C|pos) using the dataframe directly:', df[df['test_result'] == "Positive"]['has_cancer'].mean())
print('P(C|pos) using Bayes rule (and rounded results):', .905*.105 / (.905*.105 + .204*.895))

P(C|pos) using the dataframe directly: 0.34282178217821785
P(C|pos) using Bayes rule (and rounded results): 0.34230291241151994


In [53]:
print('P(~C|pos) using the dataframe directly:', (df[df['test_result'] == "Positive"]['has_cancer'] == False).mean())
print('P(~C|pos) using Bayes rule (and rounded results):', .204*.895 / (.905*.105 + .204*.895))

P(~C|pos) using the dataframe directly: 0.6571782178217822
P(~C|pos) using Bayes rule (and rounded results): 0.65769708758848


In [54]:
print('P(C|neg) using the dataframe directly:', df[df['test_result'] == "Negative"]['has_cancer'].mean())
print('P(C|neg) using Bayes rule (and rounded results):', .095*.105 / (.095*.105 + .796*.895))

P(C|neg) using the dataframe directly: 0.013770180436847104
P(C|neg) using Bayes rule (and rounded results): 0.013808235106832134


In [55]:
print('P(~C|neg) using the dataframe directly:', (df[df['test_result'] == "Negative"]['has_cancer'] == False).mean())
print('P(~C|neg) using Bayes rule (and rounded results):',  .796*.895 / (.095*.105 + .796*.895))

P(~C|neg) using the dataframe directly: 0.9862298195631529
P(~C|neg) using Bayes rule (and rounded results): 0.986191764893168


## Distributions Worth Knowing

### Discrete Distributions

##### [Binomial Distribution](https://en.wikipedia.org/wiki/Binomial_distribution)

The binomial distribution gives the probability of $k$ number of successes in $n$ independent trials, where each trial has a probability $p$ of success. Its PMF is
$$P(X=k) = {n\choose k}p^k(1-p)^{n-k}$$

Its mean is $\mu=np$ and variance is $\sigma^2=np(1-p)$.

The most common applications for a binomial distribution are coin flips, user sign-ups, and any scenario involving counting the number of successful events and the outcome is binary.

##### [Poisson Distribution](https://en.wikipedia.org/wiki/Poisson_distribution)

The Poisson distribution gives the probability of the number of events occurring within a particular fixed interval where the known, constant rate of each event's occurrence is $\lambda$. Its PMF is
$$P(X=k)=\frac{e^{-\lambda}\lambda^k}{k!}$$

Its mean is $\mu=\lambda$ and variance is $\sigma^2=\lambda$.

The most common applications of a Poisson distribution are in assessing counts over a continuous interval, such as the number of visits to a website in a certain period of time or the number of defects in a square foot of fabric. Thus, applications of the Poisson will involve a process $X$ occurring at a rate $\lambda$.

### Continuous Distributions

##### [Uniform Distribution](https://en.wikipedia.org/wiki/Continuous_uniform_distribution)

The uniform distribution assumes a constant probability of an $X$ falling between values on the interval $a$ to $b$. Its PDF is
$$f(x)=\frac{1}{b-a}$$

Its mean is $\mu=\frac{a+b}{2}$ and variance is $\sigma^2=\frac{(b-a)^2}{12}$.

The most common applications for a uniform distribution are in sampling and hypothesis testing cases.

##### [Exponential Distribution](https://en.wikipedia.org/wiki/Exponential_distribution)

The exponential distribution gives the probability of the interval length between events of a Poisson process having a set rate parameter $\lambda$. Its PDF is
$$f(x)=\lambda e^{-\lambda x}$$

Its mean is $\mu=\frac{1}{\lambda}$ and variance is $\sigma^2=\frac{1}{\lambda^2}$.

The most common applications for an exponential distribution are in wait times, such as the time until a customer makes a purchase or the time until a default in credit occurs. One of the distribution's most useful properties is the property of [memorylessness](https://en.wikipedia.org/wiki/Memorylessness).

##### [Normal Distribution](https://en.wikipedia.org/wiki/Normal_distribution)

See [Distributions/Normal Distribution.ipynb](./Distributions/Normal%20Distribution.ipynb) for details on the normal distribution.

$$f(x)=\frac{1}{\sqrt{2\pi \sigma^2}}e^{-\frac{(x-\mu)^2}{2\sigma^2}}$$

Its mean is $\mu=\mu$ and variance is $\sigma^2=\sigma^2$.

## [Markov Chains](https://en.wikipedia.org/wiki/Markov_chain)

A Markov chain (or Markov process) is a stochastic model describing a sequence of possible events in which the probability of each event depends only on the state attained in the previous event. Informally, this may be thought of as, "What happens next depends only on the state of affairs ***now***.

A Markov process is a stochastic process that satisfies the [Markov Property](https://en.wikipedia.org/wiki/Markov_property) (sometimes characterized as [memorylessness](https://en.wikipedia.org/wiki/Memorylessness)). Predictions can be made regarding future outcomes based solely on its present state and, such predictions, are just as good as the ones that could be made knowing the process's full history. Stated another way, the Markov property is such that, conditioned on the current state, the past and future states it will occupy are conditionally independent.

The probability of transitioning from state $i$ to state $j$ at any given time is given my a transition matrix, denoted by $P$.
$$P = 
\begin{pmatrix}
p_{11} & \dots & p_{1n} \\
\dots & & \dots \\
p_{m1} & \dots & p_{mn} \\
\end{pmatrix}
$$

A *recurrent state* is one where if entering the state, one will always transition back into that state eventually. In contrast, a *transient state* is one in which, if entered, there is a positive probability that upon leaving, one will never enter that state again.

A [stationary distribution](https://en.wikipedia.org/wiki/Stationary_process) for a Markov chain satisfies the following characteristic: $\pi=\pi P$, where a $P$ is a transition matrix and remains fixed following any transitions using $P$. Thus, $P$ contains the long-run proportions of the time that a process will spend in any particular state over time.